File: shared/stems.ts
/**
 * Canonical stem vocabulary — the TypeScript mirror of
 * `python/sipra_core/stems.py`.
 *
 * A test asserts the two stay in step. If you change one, change both.
 */

export type StemId = 'vocals' | 'drums' | 'bass' | 'guitar' | 'piano' | 'other';

export interface StemDefinition {
  readonly id: StemId;
  readonly label: string;
  readonly color: string;
  readonly order: number;
  readonly experimental: boolean;
  readonly note: string;
}

export const STEM_DEFINITIONS: readonly StemDefinition[] = [
  { id: 'vocals', label: 'Vocals', color: '#FF6B4A', order: 0, experimental: false, note: '' },
  { id: 'drums', label: 'Drums', color: '#FFB020', order: 1, experimental: false, note: '' },
  { id: 'bass', label: 'Bass', color: '#7C5CFF', order: 2, experimental: false, note: '' },
  {
    id: 'guitar',
    label: 'Guitar',
    color: '#2ECC71',
    order: 3,
    experimental: true,
    note: 'Separated only by the 6-stem model. Usable, but expect some bleed.',
  },
  {
    id: 'piano',
    label: 'Piano',
    color: '#35B7FF',
    order: 4,
    experimental: true,
    note:
      "The weakest source in the 6-stem model. Demucs' own documentation reports heavy " +
      'bleeding and artefacts here. Treat it as a rough guide.',
  },
  { id: 'other', label: 'Other', color: '#9AA3B2', order: 5, experimental: false, note: '' },
] as const;

export const STEM_IDS: readonly StemId[] = STEM_DEFINITIONS.map((s) => s.id);

export const STEM_BY_ID: Readonly<Record<StemId, StemDefinition>> = Object.fromEntries(
  STEM_DEFINITIONS.map((s) => [s.id, s]),
) as Record<StemId, StemDefinition>;

export const FOUR_STEM_SET: readonly StemId[] = ['vocals', 'drums', 'bass', 'other'];
export const SIX_STEM_SET: readonly StemId[] = [
  'vocals',
  'drums',
  'bass',
  'guitar',
  'piano',
  'other',
];

export function isStemId(value: unknown): value is StemId {
  return typeof value === 'string' && value in STEM_BY_ID;
}

/** Canonical display order. Unknown ids are kept, sorted, and put last. */
export function sortStems(ids: readonly string[]): string[] {
  const known: StemId[] = [];
  const unknown: string[] = [];
  for (const id of ids) {
    if (isStemId(id)) known.push(id);
    else unknown.push(id);
  }
  known.sort((a, b) => STEM_BY_ID[a].order - STEM_BY_ID[b].order);
  unknown.sort();
  return [...known, ...unknown];
}

export function stemLabel(id: string): string {
  return isStemId(id) ? STEM_BY_ID[id].label : id;
}

export function stemColor(id: string): string {
  return isStemId(id) ? STEM_BY_ID[id].color : '#9AA3B2';
}

File: python/sipra_core/stems.py
"""Canonical stem vocabulary.

The same six ids appear in the Python core, the Electron main process and
the renderer. ``shared/stems.ts`` is the TypeScript mirror of this file
and the two are kept in sync by a test.
"""

from __future__ import annotations

from dataclasses import dataclass

VOCALS = "vocals"
DRUMS = "drums"
BASS = "bass"
GUITAR = "guitar"
PIANO = "piano"
OTHER = "other"


@dataclass(frozen=True)
class StemDefinition:
    id: str
    label: str
    color: str
    order: int
    experimental: bool = False
    note: str = ""


STEM_DEFINITIONS: tuple[StemDefinition, ...] = (
    StemDefinition(VOCALS, "Vocals", "#FF6B4A", 0),
    StemDefinition(DRUMS, "Drums", "#FFB020", 1),
    StemDefinition(BASS, "Bass", "#7C5CFF", 2),
    StemDefinition(
        GUITAR,
        "Guitar",
        "#2ECC71",
        3,
        experimental=True,
        note="Separated only by the 6-stem model. Usable, but expect some bleed.",
    ),
    StemDefinition(
        PIANO,
        "Piano",
        "#35B7FF",
        4,
        experimental=True,
        note=(
            "The weakest source in the 6-stem model. Demucs' own documentation "
            "reports heavy bleeding and artefacts here. Treat it as a rough guide."
        ),
    ),
    StemDefinition(OTHER, "Other", "#9AA3B2", 5),
)

STEM_IDS: tuple[str, ...] = tuple(d.id for d in STEM_DEFINITIONS)
STEM_BY_ID: dict[str, StemDefinition] = {d.id: d for d in STEM_DEFINITIONS}

FOUR_STEM_SET: tuple[str, ...] = (VOCALS, DRUMS, BASS, OTHER)
SIX_STEM_SET: tuple[str, ...] = (VOCALS, DRUMS, BASS, GUITAR, PIANO, OTHER)


def sort_stems(ids: list[str] | tuple[str, ...]) -> list[str]:
    """Return ``ids`` in canonical display order; unknown ids go last."""
    known = [i for i in ids if i in STEM_BY_ID]
    unknown = [i for i in ids if i not in STEM_BY_ID]
    known.sort(key=lambda i: STEM_BY_ID[i].order)
    return known + sorted(unknown)


def describe(ids: list[str] | tuple[str, ...]) -> list[dict]:
    """Serialisable stem metadata for the UI."""
    return [
        {
            "id": d.id,
            "label": d.label,
            "color": d.color,
            "order": d.order,
            "experimental": d.experimental,
            "note": d.note,
        }
        for d in (STEM_BY_ID[i] for i in sort_stems(list(ids)))
    ]
